### PR TITLE
Add enhancement to rename cluster

### PR DESCRIPTION
**Pull request by knrt10**
This adds feature where user can now use renamed cluster name in their UI.

Fixes: #1653 

### Testing

- Remove old kubeconfig `rm ~/.config/Headlamp/kubeconfigs/config`
- Run app `cd app && npm run dev-only-app`
- Run backend with stateless clusters enabled `cd backend && go run ./cmd -dev -enable-dynamic-clusters`
- Go to cluster settings and there is option for user to rename their cluster

### Screenshots

Screenshot from 2024-05-13 12-49-05

Screenshot from 2024-05-13 12-49-20

### TODO

- [x] Add for stateless clusters
- [x] Add tests

### CODE REVIEW

File: frontend/src/components/App/Settings/SettingsCluster.tsx [hunk 1]
```diff
@@ -39,12 +51,23 @@ export default function SettingsCluster() {
...
+
   const classes = useStyles();
   const theme = useTheme();
 
   const history = useHistory();
   const dispatch = useDispatch();
 
+  const handleUpdateClusterName = () => {
+    try {
+      console.log('came here', newClusterName); // Call the backend function to update the cluster name
```
Comment (ashu8912) [inline frontend/src/components/App/Settings/SettingsCluster.tsx]: I think the log came here by mistake, maybe just a comment we can use here.

File: frontend/src/components/App/Settings/SettingsCluster.tsx [hunk 2]
```diff
@@ -147,12 +174,33 @@ export default function SettingsCluster() {
     });
   }
 
+  function storeNewClusterName(name: string) {
+    let actualName = name;
+    if (name === cluster) {
```
Comment (ashu8912) [inline frontend/src/components/App/Settings/SettingsCluster.tsx:192]: can you please explain the logic here, Why would we set it to empty name if the cluster name has not changed??

File: frontend/src/components/App/Layout.tsx [hunk 3]
```diff
@@ -70,7 +70,9 @@ export default function Layout({}: LayoutProps) {
    * indexDB and then sends the backend to parse it and then updates the parsed value into redux
    * store on an interval.
    * */
+  console.log('Layout.tsx: Layout()');
```
Comment (ashu8912) [inline frontend/src/components/App/Layout.tsx:73]: debug log which shouldn't be here

File: frontend/src/components/App/Layout.tsx [hunk 4]
```diff
@@ -70,7 +70,9 @@ export default function Layout({}: LayoutProps) {
...
   useEffect(() => {
+    console.log('Mounted');
```
Comment (ashu8912) [inline frontend/src/components/App/Layout.tsx:75]: same

File: frontend/src/components/App/Settings/SettingsCluster.tsx [hunk 5]
```diff
@@ -59,10 +59,23 @@ export default function SettingsCluster() {
...
       storeNewClusterName(newClusterName);
+      renameCluster(cluster || '', newClusterName, source)
+        .then(config => {
+          dispatch(setConfig(config));
+          history.push('/');
+          window.location.reload();
+        })
+        .catch((err: Error) => {
+          if (err.message === 'Not Found') {
+            // TODO: create notification with error message
```
Comment (ashu8912) [inline frontend/src/components/App/Settings/SettingsCluster.tsx:76]: Why is this a TODO?? Do you wish to do it in a different PR??

File: backend/cmd/headlamp.go [hunk 6]
```diff
@@ -1289,6 +1290,149 @@ func (c *HeadlampConfig) deleteCluster(w http.ResponseWriter, r *http.Request) {
...
+	if source == "kubeconfig" {
+		return c.kubeConfigPath, nil
+	}
+
+	return defaultKubeConfigPersistenceFile()
+}
+
+// Handler for renaming a cluster.
+//
+//nolint:funlen
```
Comment (ashu8912) [inline backend/cmd/headlamp.go:1304]: I think we can prevent this by moving the looping part of config.Contexts into a different function and letting it handle the contextName setting?? Would that be something difficult to achieve??? Does it add more complexity??

File: backend/cmd/cluster.go [hunk 7]
```diff
@@ -28,3 +28,9 @@ type ClusterReq struct {
 type KubeconfigRequest struct {
```
Comment (ashu8912) [inline backend/cmd/cluster.go:28]: Shouldn't this commit be before the frontend commits??

File: frontend/src/components/App/Settings/SettingsCluster.tsx (hunk 2 above)
Comment (knrt10) [reply frontend/src/components/App/Settings/SettingsCluster.tsx:192]: Sure. If they are equal, it means that the user hasn't changed the name of the cluster. In this case:
It sets `actualName` to an empty string. This effectively clears the new name, indicating that the user doesn't want to change the cluster name.

It calls `setNewClusterName` to update the state with the empty string, ensuring that the UI reflects this change (if applicable). 

If the user attempts to rename the cluster but provides the same name as before, it indicates that they don't want to change the name. Setting it to an empty string reflects this intention.

Regardless of whether the cluster name has changed or not, it updates the `clusterSettings` state.

File: frontend/src/components/App/Settings/SettingsCluster.tsx (hunk 5 above)
Comment (knrt10) [reply frontend/src/components/App/Settings/SettingsCluster.tsx:76]: I think this is not part of this PR, but in scope of notifications. Please refer to https://github.com/headlamp-k8s/headlamp/pull/1974/commits/c5f310e551daa7e4eac6dda5dfebf8018afbe3ac#diff-360436c42cc1c1880e31a8c726c2dc1cd6e52ed471da9443e9956ac8f7210d5aL79

File: backend/cmd/headlamp.go [hunk 8]
```diff
@@ -1056,8 +1056,9 @@ func (c *HeadlampConfig) getClusters() []Cluster {
 			Server:   context.Cluster.Server,
 			AuthType: context.AuthType(),
 			Metadata: map[string]interface{}{
-				"source":    context.SourceStr(),
-				"namespace": context.KubeContext.Namespace,
+				"source":     context.SourceStr(),
+				"namespace":  context.KubeContext.Namespace,
+				"extensions": context.KubeContext.Extensions,
```
[the remark this answers is not in the document]
Comment (knrt10) [reply backend/cmd/headlamp.go:1061]: We did not create additional extensions. It is part of k8s kubeconfig contexts. We are making direct changes to kubeconfig. Please refer to [k8s docs](https://kubernetes.io/docs/reference/config-api/kubeconfig.v1/#Context) more context.

File: backend/cmd/headlamp.go (hunk 6 above)
Comment (knrt10) [reply backend/cmd/headlamp.go:1304]: It might not be easy because we have logging which should be part of k8s handler which creates additional lines. Even if we move some logic to other function it is not possible with clear code to have it under 60 with logging involved.

Comment (illume) [reply backend/cmd/headlamp.go:1304]: I agree with Ashu and the linter about this function being too long. It will be easier to add tests then as well.

**Review by joaquimrocha — CHANGES_REQUESTED** [review]

**Comment by joaquimrocha** [comment]
So I swear I had sent a message here with some screenshots showing some problems but I don't see the message now...
Anyway, I pushed some changes to make the UI hopefully nicer + restrict it to electron only (this is important!), and I have also verified that some of the problems I had which made me think we shouldn't add this to the 0.24.0 release were in a past iteration of this branch.
Now things seem to work fine, though it looks like if I switch between the default cluster name and the new one, in the settings it shows me the previous name (not sure if this is autocomplete getting in the way).

Anyway, here is how it looks now (notice how the error status no longer extends the input till the end of the space):
Screenshot

**Comment by joaquimrocha** [comment]
@illume What do you think? Should try to add it for 0.24.0 or should we move it (keep it now, since I had changed the milestone when I was in the outdated branch) to 0.25.0?

File: frontend/src/i18n/locales/pt/translation_old.json
Comment (joaquimrocha) [inline frontend/src/i18n/locales/pt/translation_old.json:1]: *_old.json files shouldn't be checked into the repo.

**Review by joaquimrocha — DISMISSED** [review]

**Review by illume — APPROVED** [review]
👍 thanks

**Comment by illume** [comment]
@knrt10 

Seems the text is wrong after I change it, then change it back.

- click 'minikube' cluster and cluster settings
- change 'minikube' to 'mycluster'
- change back to 'minikube'
- click cluster, and cluster settings... it says 'mycluster' still in the input text.


See in the minikube cluster but the text has the old 'mycluster' there. It should be 'minikube' or empty.
image

**Review by illume — DISMISSED** [review]
I had a look around for the bug a little bit, but couldn't find it.

Probably we should add some tests, and perhaps refactor SettingsCluster a bit.

**Comment by illume** [comment]
Maybe we should add this flag to dev mode? Or to the app?

```
diff --git a/Makefile b/Makefile
index 2d426825..102fd637 100644
--- a/Makefile
+++ b/Makefile
@@ -86,10 +86,10 @@ run-backend:
        @echo "**** Warning: Running with Helm and dynamic-clusters endpoints enabled. ****"

 ifeq ($(UNIXSHELL),true)
-       HEADLAMP_BACKEND_TOKEN=headlamp HEADLAMP_CONFIG_ENABLE_HELM=true HEADLAMP_CONFIG_ENABLE_DYNAMIC_CLUSTERS=true ./backend/headlamp-server -dev -proxy-urls https://artifacthub.io/*
+       HEADLAMP_BACKEND_TOKEN=headlamp HEADLAMP_CONFIG_ENABLE_HELM=true HEADLAMP_CONFIG_ENABLE_DYNAMIC_CLUSTERS=true ./backend/headlamp-server -dev -proxy-urls https://artifacthub.io/* -enable-dynamic-clusters
 else
        @echo "**** Running on Windows without bash or zsh. ****"
-       @cmd /c "set HEADLAMP_BACKEND_TOKEN=headlamp&& set HEADLAMP_CONFIG_ENABLE_HELM=true&& set HEADLAMP_CONFIG_ENABLE_DYNAMIC_CLUSTERS=true&& backend\headlamp-server -dev -proxy-urls https://artifacthub.io/*"
+       @cmd /c "set HEADLAMP_BACKEND_TOKEN=headlamp&& set HEADLAMP_CONFIG_ENABLE_HELM=true&& set HEADLAMP_CONFIG_ENABLE_DYNAMIC_CLUSTERS=true&& backend\headlamp-server -dev -proxy-urls https://artifacthub.io/*" -enable-dynamic-clusters
 endif

 run-frontend:
```

**Comment by knrt10** [comment]
@illume fixed, can you please test again. Thanks

**Comment by illume** [comment]
I'm having an issue where it's not changing...

image

**Comment by illume** [comment]
Ok... it works after I clear my localstorage. I think when it was broken something bad got in there.

**Review by skoeva — APPROVED** [review]
Works great on my end, thanks for this! 🙌
